Replace bzero() with memset()

nelsonuhan opened the pull request:
I replaced `bzero()` in `kkt.c` with `memset()`. 

I did this in order to get pyglpk to compile on Windows with MSVC 2015. Apparently, `bzero()` is deprecated - from its (Linux) man page:

> This function is deprecated (marked as LEGACY in POSIX.1-2001): use memset(3) in new programs.  POSIX.1-2008 removes the specification of bzero().

nelsonuhan committed:
Replace bzero() with memset()
bzero() is deprecated (removed in POSIX.1-2008), doesn’t work with MSVC 2015.

bradfordboyle commented on src/kkt.c:
```diff
@@ -34,7 +34,7 @@ static void KKT_dealloc(KKTObject *self) {
   KKTObject *k = (KKTObject*)PyObject_New(KKTObject, &KKTType);
   if (k==NULL) return k;
-  bzero((void*)(&(k->kkt)), sizeof(pyglpk_kkt_t));
+  memset((void*)(&(k->kkt)), 0, sizeof(pyglpk_kkt_t));
```
Small nitpick: The cast to `void *` is not [needed](http://stackoverflow.com/a/16534673) and because of operator precedence the parentheses around `k->kkt` are [unneeded](https://en.wikipedia.org/wiki/Operators_in_C_and_C%2B%2B#Operator_precedence). I think

``` c
memset(&k->kkt, 0, ...
```

is a bit easier to parse.

nelsonuhan replied:
Thanks for the feedback - I made the change.

bradfordboyle reviewed (commented):
This is a relatively straight-forward change. However, this line of code is not currently being covered by the current sets of tests. Would you be able to add a test to `SimpleSolverTest` to cover this section of the code?

nelsonuhan committed:
Small edit to make code easier to parse

nelsonuhan committed:
bug fix: Fix attribute names defined in KKT_getset

nelsonuhan committed:
Add tests for KKT check with solutions from simplex and exact solvers

nelsonuhan committed:
Add simple tests for integer solver and KKT check using solution from integer solver

nelsonuhan committed:
Revise simple integer tests to use IP with unique optimal solution

nelsonuhan commented:
I added some tests to cover:
- `pyglpk_kkt_check()` using solutions from the simplex and exact solvers,
- `integer()` solving a simple integer program,
- `pyglpk_int_check()` using a solution from the integer solver.

I also fixed a small bug in the attribute names defined in `KKT_getset`.

bradfordboyle commented on src/kkt.c:
```diff
@@ -184,9 +184,9 @@ static PyGetSetDef KKT_getset[] = {
    "'H', high, 'M', medium, 'L', low, or '?' wrong or infeasible."},
 
-  {"de_ae_row", (getter)KKT_de_ae_col, (setter)NULL,
+  {"de_ae_col", (getter)KKT_de_ae_col, (setter)NULL,
```
I'm a little hesitant to merge this pull request with this change for two main reasons:
1. this technically breaks backwards compatibility (existing code that works and uses `de_ae_row` will stop working)
2. there are places in the documentation that need to be updated to reflect this change.

Everything else can be merged to master and the version number can be bumped to v0.4.3. I'll start a v0.5.0-develop branch and we can merge this change there.

nelsonuhan replied:
Good point. I changed those attribute names back, and fixed the tests accordingly. 

I'll submit a new PR with the name changes after I comb through and figure out what parts of the documentation need to be updated.

nelsonuhan committed:
Revert back to de_ae_row and de_re_row in KKT_getset
Fix tests accordingly

nelsonuhan committed:
Fix typos with comments in KKT tests
